pkg/controllers/state/state.go

```diff
@@ -0,0 +1,241 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"knative.dev/pkg/logging"
+
+	"golang.org/x/time/rate"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	podutils "github.com/aws/karpenter/pkg/utils/pod"
+	"github.com/aws/karpenter/pkg/utils/resources"
+)
+
+// Cluster maintains cluster state that is often needed but expensive to compute.
+type Cluster struct {
+	ctx         context.Context
+	rateLimiter *rate.Limiter
+	kubeClient  client.Client
+
+	// Pod Specific Tracking
+	knownPods        sync.Map // mapping of pod namespaced name to struct{} used to track if we've seen a pod before
+	antiAffinityPods sync.Map // mapping of pod namespaced name to *v1.Pod of pods that have required anti affinities
+
+	// Node Status & Pod -> Node Binding
+	mu       sync.RWMutex
+	nodes    map[string]*Node                // node name -> node
+	bindings map[types.NamespacedName]string // pod namespaced named -> pod
+}
+
+func NewCluster(ctx context.Context, client client.Client) *Cluster {
+	s := &Cluster{
+		ctx:         ctx,
+		kubeClient:  client,
+		rateLimiter: rate.NewLimiter(rate.Every(10*time.Second), 1),
+		nodes:       map[string]*Node{},
+		bindings:    map[types.NamespacedName]string{},
+	}
+	return s
+}
+
+// Node is a cached version of a node in the cluster that maintains state which is expensive to compute every time it's
+// needed.  This currently contains node utilization across all of the allocatable resources, but will soon be used to
+// compute topology information.
+type Node struct {
+	Node        *v1.Node
```

Comment:
Thoughts on making it the same thing 
```suggestion
	*v1.Node
```

Reply:
Embedding usually feels wrong to me, except for some rare cases.